### PR TITLE
Make Tools own model, add ToolKit Concept

### PR DESCRIPTION
**Pull request by agola11**
Follow-up of @hinthornw's PR:

- Migrate the Tool abstraction to a separate file (`BaseTool`).
- `Tool` implementation of `BaseTool` takes in function and coroutine to more easily maintain backwards compatibility
- Add a Toolkit abstraction that can own the generation of tools around a shared concept or state

### CODE REVIEW

**Review by hwchase17 — COMMENTED** [review]
my two big comments:

im not sure we need toolkits for google search/bing/wolfram alpha cause they only have one tool. would prefer not to overdo stuff, keep it light and basic until we have more certainty

tools loded in load_tools.py should use the Tool classes, right?

File: Makefile [hunk 1]
```diff
@@ -32,9 +32,6 @@ lint:
 test:
 	poetry run pytest tests/unit_tests
 
-tests:
```
Comment (agola11) [inline Makefile:35]: @hwchase17 -- can move this to another PR but not sure why there is both `make test and make tests`

Comment (hwchase17) [reply Makefile:35]: someone added this because its more canonical, dont see harm in having both

Comment (agola11) [reply Makefile]: reverted
